internal/pn_counter.go

```diff
@@ -0,0 +1,253 @@
+	defer c.targetSelectionMutex.RUnlock()
+	_, ok := excludedAddresses[c.currentTargetReplicaAddress]
+	if c.currentTargetReplicaAddress != nil && !ok {
+		return c.currentTargetReplicaAddress, nil
+	}
+	var err error
+	_, ok = excludedAddresses[c.currentTargetReplicaAddress]
+	if c.currentTargetReplicaAddress == nil || ok {
+		c.targetSelectionMutex.RUnlock()
+		c.targetSelectionMutex.Lock()
+		c.currentTargetReplicaAddress, err = c.chooseTargetReplica(excludedAddresses)
+		c.targetSelectionMutex.Unlock()
+		c.targetSelectionMutex.RLock()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return c.currentTargetReplicaAddress, nil
+}
+
+func (c *PNCounterProxy) invokeGetInternal(excludedAddresses map[IAddress]struct{}, lastError error,
+	target IAddress) (response *ClientMessage, err error) {
+	if target == nil {
+		if lastError != nil {
+			err = lastError
+		} else {
+			err = NewHazelcastNoDataMemberInClusterError("cannot invoke operations on a CRDT because the cluster does not contain any data members", nil)
+		}
+		return
+	}
+	request := PNCounterGetEncodeRequest(c.name, (*VectorClock)(atomic.LoadPointer(&c.observedClock)).EntrySet(), target.(*Address))
+	response, lastError = c.invokeOnAddress(request, target.(*Address))
+	if lastError != nil {
+		log.Printf("error occurred while invoking operation on target %v, choosing different target", target)
+		excludedAddresses[target] = struct{}{}
+		newTarget, err := c.getCRDTOperationTarget(excludedAddresses)
+		if err != nil {
+			return nil, err
+		}
+		return c.invokeGetInternal(excludedAddresses, lastError, newTarget)
+	}
+	return response, nil
+
+}
+
+func (c *PNCounterProxy) invokeAddInternal(delta int64, getBeforeUpdate bool, excludedAddresses map[IAddress]struct{}, lastError error,
+	target IAddress) (response *ClientMessage, err error) {
+	if target == nil {
+		if lastError != nil {
+			err = lastError
+		} else {
+			err = NewHazelcastNoDataMemberInClusterError("cannot invoke operations on a CRDT because the cluster does not contain any data members", nil)
+		}
+		return
+	}
+	request := PNCounterAddEncodeRequest(c.name, delta, getBeforeUpdate, (*VectorClock)(atomic.LoadPointer(&c.observedClock)).EntrySet(), target.(*Address))
+	response, lastError = c.invokeOnAddress(request, target.(*Address))
+	if lastError != nil {
+		log.Printf("unable to provide session guarantees when sending operations to %v, choosing different target", target)
+		excludedAddresses[target] = struct{}{}
+		newTarget, err := c.getCRDTOperationTarget(excludedAddresses)
+		if err != nil {
+			return nil, err
+		}
+		return c.invokeAddInternal(delta, getBeforeUpdate, excludedAddresses, lastError, newTarget)
+	}
+	return response, nil
+}
+
+func (c *PNCounterProxy) chooseTargetReplica(excludedAddresses map[IAddress]struct{}) (IAddress, error) {
+	replicaAddresses, err := c.getReplicaAddresses(excludedAddresses)
+	if err != nil || len(replicaAddresses) == 0 {
+		return nil, err
+	}
+	rand.Seed(time.Now().UnixNano())
```

Comment:
If seed is called once, `rand.Intn` returns always same result so I preferred this way to be able to get different random values.

Reply:
I think it should not be that way. When I run https://play.golang.org/p/Vsoo4PBTlVe , it gives another random integer at each line.

Reply:
The thing confusing me is that when you run this code again, it prints same results.

Reply:
Because it is an online playground, probably `time.Now()` call returns a cached value within specific intervals. So when you run the same sample within a short amount of time, the seed stays the same. It is expected to get the same set of random numbers then.